base/abstractarray.jl

```diff
@@ -1234,11 +1234,20 @@ oneunit(x::AbstractMatrix{T}) where {T} = _one(oneunit(T), x)
 # While the definitions for IndexLinear are all simple enough to inline on their
 # own, IndexCartesian's CartesianIndices is more complicated and requires explicit
 # inlining.
-function iterate(A::AbstractArray, state=(eachindex(A),))
+_iterate_starting_state(A) = _iterate_starting_state(A, IndexStyle(A))
+_iterate_starting_state(A, ::IndexLinear) = firstindex(A)
+_iterate_starting_state(A, ::IndexStyle) = (eachindex(A),)
+iterate(A::AbstractArray, state = _iterate_starting_state(A)) = _iterate(A, state)
+function _iterate(A::AbstractArray, state::Tuple)
```

Comment:
```suggestion
iterate(A::AbstractArray) = _iterate_starting_state(A)
function iterate(A::AbstractArray, state::Tuple)
```

I think this would be simpler and more straightforward, no?  Or is there a reason why you structured it this way?

Reply:
I had added an extra level of indirection to avoid method ambiguities in `iterate`

Reply:
Ah of course, yes.  Perhaps we could use a better name than `_iterate`, though?